Marlin/src/HAL/STM32/tft/tft_spi.cpp

```diff
@@ -74,31 +64,50 @@ void TFT_SPI::Init() {
 
   pinmap_pinout(digitalPinToPinName(TFT_SCK_PIN), PinMap_SPI_SCLK);
   pinmap_pinout(digitalPinToPinName(TFT_MOSI_PIN), PinMap_SPI_MOSI);
-  #if PIN_EXISTS(TFT_MISO) && (TFT_MISO_PIN != TFT_MOSI_PIN)
-    pinmap_pinout(digitalPinToPinName(TFT_MISO_PIN), PinMap_SPI_MISO);
+  #if PIN_EXISTS(TFT_MISO)
+    if (TFT_MISO_PIN != TFT_MOSI_PIN)
+      pinmap_pinout(digitalPinToPinName(TFT_MISO_PIN), PinMap_SPI_MISO);
   #endif
   pin_PullConfig(get_GPIO_Port(STM_PORT(digitalPinToPinName(TFT_SCK_PIN))), STM_LL_GPIO_PIN(digitalPinToPinName(TFT_SCK_PIN)), GPIO_PULLDOWN);
 
   #ifdef SPI1_BASE
     if (SPIx.Instance == SPI1) {
       __HAL_RCC_SPI1_CLK_ENABLE();
-      __HAL_RCC_DMA1_CLK_ENABLE();
+      #ifdef STM32F1xx
+        __HAL_RCC_DMA1_CLK_ENABLE();
+        DMAtx.Instance = DMA1_Channel3;
+      #elif defined(STM32F4xx)
+        __HAL_RCC_DMA2_CLK_ENABLE();
+        DMAtx.Instance = DMA2_Stream3;
+        DMAtx.Init.Channel = DMA_CHANNEL_3;
+      #endif
```

Comment:
This will be a problem for MCUs other than F1 and F4.

We must remember that currently this HAL is used for F0, F1, F4, F7, and soon H7 MCUs.

Anywhere we must behave differently on each MCU family, we should add something like this to ensure it is not used with unimplemented hardware.
```
#else
  #error "Not implemented for this MCU"
#endif
```

You may be able to avoid doing this in many places with some macros, with something like this, although you only have three places right now so it might not be worth the trouble.

```
#ifdef STM32F1xx
  #define SETUP_DMA(dest, dma, channel) do { \
    __HAL_RCC_DMA##dma##_CLK_ENABLE(); \
    (dest).Instance = DMA##dma##_Channel##channel; \
  } while(0)
#elif defined(STM32F4xx)
  #define SETUP_DMA(dest, dma, channel) do { \
    __HAL_RCC_DMA##dma##_CLK_ENABLE(); \
    (dest).Instance = DMA##dma##_Stream##channel; \
    (dest).Init.Channel = DMA_CHANNEL_##channel; \
  } while(0)
#else
  #error "Not implemented for this MCU"
#endif

Reply:
We can do it, but I don't think it worth do it now, for 2 reasons: all those SPI code will move to MarlinSPI soon, and enabling a TFT for others MCU may break code before that. Even the board needs specific pins set, and only F4 and F1 boards have it set now. So the build will stop before get this point.

Reply:
I think you should put an `#error` at least once in each file which will cause problems. This is an aid for any future development, to make sure files that need to be updated cannot be missed.

Reply:
Maybe it's matter for a sanity check: TFT will only work on F1 and F4. The #error for dma/spi can be added in the MarlinSPI, I guess.